src/clj_xpath/lib.clj

```diff
@@ -0,0 +1,587 @@
+  [bytes & [opts]]
+  (with-open [istr (ByteArrayInputStream. bytes)]
+    (input-stream->dom istr opts)))
+
+(defmulti  xml->doc
+  "Convert various forms of XML into a Document.  Supported forms:
+
+    String
+    byte array
+    Input Stream
+    org.w3c.dom.Document
+    org.w3c.dom.Node
+  "
+  (fn [thing & [opts]] (class thing)))
+
+(defmethod xml->doc String               [thing & [opts]] (xml-bytes->dom (.getBytes ^String thing ^String (:default-encoding opts "YTF-8")) opts))
+(defmethod xml->doc (Class/forName "[B") [thing & [opts]] (xml-bytes->dom thing opts))
+(defmethod xml->doc InputStream          [thing & [opts]] (input-stream->dom thing opts))
+(defmethod xml->doc org.w3c.dom.Document [thing & [opts]] thing)
+(defmethod xml->doc Node                 [thing & [opts]] thing)
+(defmethod xml->doc :default             [thing & [opts]]
+  (throwf "Error, don't know how to build a doc out of '%s' of class %s" thing (class thing)))
+
+
+(defn attrs
+  "Extract the attributes from the node."
+  [^Node nodeattrs]
+  (if-let [the-attrs (.getAttributes nodeattrs)]
+    (loop [[^Node node & nodes] (node-list->seq (.getAttributes nodeattrs))
+           res {}]
+      (if node
+        (recur nodes (assoc res (keyword (.getNodeName node)) (.getTextContent node)))
+        res))
+    nil))
+
+(defn text
+  "Accessor for text content from the node."
+  [#^Node node]
+  (.getTextContent node))
+
+(defn node-name
+  "Accessor for the node's name."
+  [#^Node node]
+  (keyword (.getNodeName node)))
+
+
+(defn node->map
+  "Create a logical map out of the Node's properties:
+    :node      the-node
+    :tag       tag name of node
+    :attrs     map of the node's attributes
+    :text      the text of the node
+    :children  a lazy sequence of the node's children.
+  "
+  [#^Node node]
+  (let [lazy-children (fn [#^Node n] (delay
+                                      (map node->map
+                                           (node-list->seq (.getChildNodes n)))))
+        m  {:node node
+            :tag   (node-name node)
+            :attrs (attrs node)
+            :text  (text node)}
+        m   (if (.hasChildNodes node)
+              (assoc m :children (lazy-children node))
+              m)]
+    m))
+
+
+(defmulti xp:compile
+  "Compile an XPath expression.  If the argument is already a compiled XPath expression, it is returned as-is."
+  (fn [xp-compiler xpexpr] (class xpexpr)))
+
+(defmethod xp:compile String          [xp-compiler xpexpr] (.compile ^XPath xp-compiler xpexpr))
+(defmethod xp:compile XPathExpression [xp-compiler xpexpr] xpexpr)
+(defmethod xp:compile :default        [xp-compiler xpexpr]
+  (throwf "xp:compile: don't know how to compile xpath expr of type:%s '%s'" (class xpexpr) xpexpr))
+
+
+(defmulti $x
+  "Perform an xpath query on the given XML document which may be a String, byte array, or InputStream.
+  See xml->doc, and xp:compile."
+  (fn [xp-compiler xp xml-thing & [opts]] (class xml-thing)))
```

Comment:
Instead of xp-compiler being passed as a positional parameter, should it be absorbed into the `opts` map? It seems both of those parameters need to end up in the same place. `$x` and the `xmlnsmap` functions would need to be updated.

Having a single `opts` map to pass to different functions would create a common protocol of sorts between functions so I think the library would be easier to modify and extend without breaking things in the future. For example, it would enable convenience macros like korma’s `defentity`.